source/_posts/2017-11-15-code-generation-with-sourcery.md

```diff
@@ -0,0 +1,133 @@
+
+```swift
+final class AppUser: Model, Timestampable, SoftDeletable {
+    struct Keys {
+        static let email = "email"
+        static let name = "name"
+        static let password = "password"
+        static let passwordRepeat = "passwordRepeat"
+        static let companyId = "companyId"
+    }
+    
+    public let storage = Storage()
+    
+    public var email: String
+    public var name: String?
+    public var password: String
+    public var companyId: Identifier
+    
+    public init(
+        email: String,
+        name: String?,
+        password: String,
+        companyId: Identifier
+    ) {
+        self.email = email
+        self.name = name
+        self.password = password
+        self.companyId = companyId
+    }
+
+    public init(row: Row) throws {
+        name = try row.get(Keys.name)
+        email = try row.get(Keys.email)
+        password = try row.get(Keys.password)
+        companyId = try row.get(Keys.companyId)
+    }
+  
+    internal enum JSONKeys: String {
+        case name
+        case email
+        case companyId
+    }
+}
+
+extension AppUser: JSONRepresentable {
+    public func makeJSON() throws -> JSON {
+        var json = JSON()
+        try json.set(idKey, id)
+        try json.set(JSONKeys.name.rawValue, name)
+        try json.set(JSONKeys.email.rawValue, email)
+        try json.set(JSONKeys.companyId.rawValue, companyId)
+        return json
+    }
+}
+
+extension AppUser: NodeRepresentable {
+    public func makeNode(in context: Context?) throws -> Node {
+        var node = Node([:])
+        try node.set(idKey, id)
+        try node.set(Keys.name, name)
+        try node.set(Keys.email, email)
+        try node.set(Keys.companyId, companyId)
+        try node.set(AppUser.createdAtKey, createdAt)
+        try node.set(AppUser.updatedAtKey, updatedAt)
+        try node.set("company", company.get())
+        return node
+    }
+}
+
+extension AppUser: Preparation {
+    public static func prepare(_ database: Database) throws {
+        try database.create(self) { user in
+            user.id()
+            user.string(Keys.email, unique: true)
+            user.string(Keys.name, optional: true)
+            user.string(Keys.password)
+            user.string(Keys.companyId)
+        }
+        
+        
+        try database.index("email", for: AppUser.self)
+        try database.index(Keys.companyId, for: AppUser.self)
+    }
+    
+    public static func revert(_ database: Database) throws {
+        try database.delete(self)
+    }
+}
+
+extension AppUser {
+    public func makeRow() throws -> Row {
+        var row = Row()
+        try row.set(Keys.email, email)
+        try row.set(Keys.name, name)
+        try row.set(Keys.password, password)
+        try row.set(Keys.companyId, companyId)
+        return row
+    }
+}
+```
+
```

Comment:
I agree with the TODO list, maybe a small breakdown on what the different pieces are doing and why we need them.